src/sw.ts

```diff
@@ -177,13 +186,93 @@ function isSwAssetRequest (event: FetchEvent): boolean {
   return isActualSwAsset
 }
 
+/**
+ * Set the expires header on a response object to a timestamp based on the passed ttl interval
+ */
+function setExpiresHeader (response: Response, ttlSeconds: number = 3600): void {
+  const expirationTime = new Date(Date.now() + ttlSeconds * 1000)
+
+  response.headers.set('Expires', expirationTime.toUTCString())
+}
+
+/**
+ * Checks whether a cached response object has expired by looking at the expires header
+ * Note that this ignores the Cache-Control header since the expires header is set by us
+ */
+function hasExpired (response: Response): boolean {
+  const expiresHeader = response.headers.get('Expires')
+
+  if (expiresHeader == null) {
+    return false
+  }
+
+  const expires = new Date(expiresHeader)
+  const now = new Date()
+
+  if (expires < now) {
+    return true
+  }
+
+  return false
+}
+
+async function getResponseFromCacheOrFetch (event: FetchEvent): Promise<Response> {
+  const { protocol } = getSubdomainParts(event.request.url)
+  const url = new URL(event.request.url)
+  const isMutable = protocol === 'ipns'
+  const cacheKey = `${event.request.url}-${event.request.headers.get('Accept') ?? ''}`
+  trace('helia-sw: cache key: %s', cacheKey)
+  const cache = await caches.open(isMutable ? MUTABLE_CACHE : IMMUTABLE_CACHE)
+  const cachedResponse = await cache.match(cacheKey)
+
+  if ((cachedResponse != null) && !hasExpired(cachedResponse)) {
+  // If there is an entry in the cache for event.request,
+  // then response will be defined and we can just return it.
+    log('helia-ws: cached response HIT for %s (expires: %s) %o', cacheKey, cachedResponse.headers.get('Expires'), cachedResponse)
+
+    trace('helia-ws: updating cache for %s in the background', cacheKey)
+    // 👇 update cache in the background (don't await)
+    fetchHandler({ path: url.pathname, request: event.request })
+      .then(async response => storeReponseInCache({ response, isMutable, cache, cacheKey }))
+      .catch(err => {
+        err('helia-ws: failed updating response in cache for %s in the background', cacheKey, err)
+      })
+
+    return cachedResponse
+  }
+
+  // 👇 fetch because no cached response was found
+  const response = await fetchHandler({ path: url.pathname, request: event.request })
+
+  void storeReponseInCache({ response, isMutable, cache, cacheKey }).catch(err => {
+    err('helia-ws: failed storing response in cache for %s', cacheKey, err)
+  })
+
+  return response
```

Comment:
I feel like we could combine these so we don't have two potential places where cache storing happens.

Reply:
looking at the code more locally, we don't need to fetch from the network at all on cache-miss.  Since we're currently fighting rate limits from the trustless gateway, we could just not do network requests. The likelihood of something updating on IPFS is very low, whereas stale-while-revalidate is good for frequently updated content that you want a quick response for.

Reply:
resolved in #109